cobbler/utils/threading.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module is responsible for managing the custom common threading logic Cobbler has.
+"""
+
+import logging
+import pathlib
+import time
+from threading import Thread
+
+from cobbler import enums
+from cobbler import utils
+
+
+class CobblerThread(Thread):
+    """
+    This is a custom thread that has a custom logger as well as logic to execute Cobbler triggers.
+    """
+
+    def __init__(self, event_id: str, remote, options: dict, task_name: str, api):
+        """
+        This constructor creates a Cobbler thread which then may be run by calling ``run()``.
+
+        :param event_id: The event-id which is associated with this thread. Also used as thread name
+        :param remote: The Cobbler remote object to execute actions with.
+        :param options: Additional options which can be passed into the Thread.
+        :param task_name: The high level task name which is used to trigger pre- and post-task triggers
+        :param api: The Cobbler api object to resolve information with.
+        """
+        Thread.__init__(self, name=event_id)
+        self.event_id = event_id
+        self.remote = remote
+        self.logger = logging.getLogger()
+        self.__setup_logger()
+        if options is None:
+            options = {}
+        self.options = options
+        self.task_name = task_name
+        self.api = api
+
+    def __setup_logger(self):
+        """
+        Utility function that will setup the Python logger for the tasks in a special directory.
+        """
+        filename = pathlib.Path("/var/log/cobbler/tasks") / f"{self.event_id}.log"
+        task_log_handler = logging.FileHandler(str(filename), encoding="utf-8")
+        task_log_formatter = logging.Formatter(
+            "[%(threadName)s] %(asctime)s - %(levelname)s | %(message)s"
+        )
+        task_log_handler.setFormatter(task_log_formatter)
+        self.logger.setLevel(logging.INFO)
+        self.logger.addHandler(task_log_handler)
+
+    def on_done(self):
+        """
+        This stub is needed to satisfy the Python inheritance chain.
+        """
+
+    def run(self):
+        """
+        Run the thread.
+
+        :return: The return code of the action. This may possibly a boolean or a Linux return code.
+        """
+        self.logger.info("start_task(%s); event_id(%s)", self.task_name, self.event_id)
+        time.sleep(1)
+        try:
+            if utils.run_triggers(
+                self.api,
+                None,
+                f"/var/lib/cobbler/triggers/task/{self.task_name}/pre/*",
+                self.options,
+            ):
+                self.remote._set_task_state(
+                    self, self.event_id, enums.EventStatus.FAILED
+                )
+                return False
+            rc = self._run(self)
```

Comment:
Where is `_run()` defined? Any advantage over `super().run()`?

Reply:
Run is assigned in the method that builds the threads. I believe we could change that mechanism to be more straight forward but I haven't dared to touch it. Since you asked me to fix it I can do it now instead of later. It is correct that we need to fix it.

Reply:
Answered above